class-bookrestcontroller.php

```diff
@@ -0,0 +1,213 @@
+<?php
+
+class BookRestController extends WP_REST_Controller {
+	private $meta_key;
+	private $cap;
+	public function __construct() {
+		$this->namespace = '/bookAPI/v1';
+		$this->meta_key  = 'myBookApp';
+		$this->cap       = current_user_can( 'edit_posts' );
+	}
+	/**
+	 * Registers the CRUD routes for GET , PUT, PATCH and DELETE
+	 * @return void
+	 */
+	public function register_routes() {
+		//get all items
+		register_rest_route(
+			$this->namespace,
+			//notice, /books/post_id is the route, books plural, the ones after this will be at singular
+			'/books/(?P<id>\d+)',
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_items' ),
+				'permission_callback' => function () {
+					return true;
+				},
+			)
+		);
+		//since the data is kept in a serialized array, put method used on create
+		register_rest_route(
+			$this->namespace,
+			'/book/(?P<id>\d+)',
+			array(
+				array(
+					'methods'             => 'PUT',
+					'callback'            => array( $this, 'create_item' ),
+					'permission_callback' => function () {
+						return $this->cap;
+					},
+				),
+			)
+		);
+		//patch for edit, not all fields must be sent
+		register_rest_route(
+			$this->namespace,
+			'/book/(?P<id>\d+)',
+			array(
+				'methods'             => 'PATCH',
+				'callback'            => array( $this, 'update_item' ),
+				'permission_callback' => function () {
+					return $this->cap;
+				},
+			)
+		);
+		//delete item
+		register_rest_route(
+			$this->namespace,
+			//use id and post_id as request parameters since both are needed in the current format
+			'/book/(?P<id>\d+)&(?P<post_id>\d+)',
+			array(
+				'methods'             => 'DELETE',
+				'callback'            => array( $this, 'delete_item' ),
+				'permission_callback' => function () {
+					return $this->cap;
+				},
+			)
+		);
+	}
+	/**
+	 * Retrieves all items from the post meta in JSON format<br>
+	 *
+	 * @param array $request MUST have the key 'id'
+	 *
+	 * @return false|string The data in format JSON or FALSE if bad request
+	 */
+	public function get_items( $request ) {
+		if ( ! $this->validate_int( $request['id'] ) ) {
+			http_response_code(400);
+			return json_encode( array( 'success' => false ) );
+		}
+		$data = get_post_meta( (int) $request['id'], $this->meta_key, true );
+		if ( $data !== false ) {
+			$data = unserialize( $data, array( 'allowed_classes' => true ) );
+		}
+		return json_encode( $data );
+	}
+	/**
+	 * Adds an item into the serialized array<br>
+	 * If the item with the specified id already exists it will be updated
```

Comment:
we should not update items on this method. we only create/add new entries in DB on this method

Reply:
updating shouldn't happen but it's a side effect since I didn't check if the item already exists